### PR TITLE
feat(content): moving fungal logic to external settings & dangerous_fungi mod

### PR DESCRIPTION
**Pull request by 4nonch**
### Required

- [x] I wrote the PR title in [conventional commit format](https://docs.cataclysmbn.org/en/contribute/changelog_guidelines/).
- [x] I ran the [code formatter](https://docs.cataclysmbn.org/en/contribute/contributing/#code-style).
- [x] I linked any relevant issues using [github keyword syntax](https://docs.cataclysmbn.org/en/contribute/contributing/#pull-request-notes) so it can be closed automatically.
- [x] I have [committed my changes to new branch that isn't `main`](https://docs.cataclysmbn.org/en/contribute/contributing/#make-your-changes) so it won't cause conflict when updating `main` branch later.

## Purpose of change
Mycus activity were heavily restricted in game. I tried to add support for mod that will return old fungal behavior while not affecting the main gameplay without it. 

## Describe the solution
- Added FUNGAL_ADVANCED attack function to calculate spore chance seemly like it was in 0.D
- MON_FUNGALOID_YOUNG_ALLOWED external option to enable/disable spawn of fungal stalk from spore
- FUNGUS_SPREAD_ON_FLAT_TILES_ALLOWED external option to allow FLAT tile infestation
- MON_FUNGALOID_YOUNG_SPAWN_BASE_RATE and MON_FUNGALOID_YOUNG_SPAWN_BUBBLE_CREATURES_DIVIDER external options to tweak chance of fungal stalk being spawned.
- Formula to spawn fungal stalk: 1 / (MON_FUNGALOID_YOUNG_SPAWN_BASE_RATE + num_creatures_in_bubble / MON_FUNGALOID_YOUNG_SPAWN_BUBBLE_CREATURES_DIVIDER)
- FUNGUS_SPORE_CHANCE external option to specify basic spawn chance for spore (single/in radius) for character actions, bashing infested furniture and terrain, FUNGUS death function, FUNGUS_ADVANCED attack function
- FUNGUS_ADVANCED_CREATURES_THRESHOLD external option to proportionally reduce chance of spore spawn as creature count inside a bubble will pass the threshold
- FUNGUS_SPORE_CREATURES_THRESHOLD external option to globally stop producing spores with fungal_effects::fungalize as creature count inside a bubble will pass the threshold

## Describe alternatives you've considered
Affecting source code directly. This not an option, i'll tried to leave the things like they be.

## Testing
Works locally.

## Additional context
None.

### CODE REVIEW

File: src/monattack.h [hunk 1]
```diff
@@ -33,15 +33,19 @@ bool grow_vine( monster *z );
...
-bool fungus_inject( monster *z );     // Directly inject the spores
-bool fungus_bristle( monster *z );    // Fungal hedgerow rake & inject
-bool fungus_growth( monster *z );     // Sporeling -> fungal creature
-bool fungus_sprout( monster *z );     // Grow fungal walls
-bool fungus_fortify( monster *z );    // Grow fungal hedgerows
+bool fungus_inject( monster *z );      // Directly inject the spores
+bool fungus_bristle( monster *z );     // Fungal hedgerow rake & inject
+bool fungus_growth( monster *z );      // Sporeling -> fungal creature
+bool fungus_sprout( monster *z );      // Grow fungal walls
+bool fungus_fortify( monster *z );     // Grow fungal hedgerows
```
Comment (scarf005) [inline src/monattack.h:48]: what are these spaces for?

File: src/cached_options.h
Comment (scarf005) [inline src/cached_options.h:1]: might make sense to store these options into struct since they're strongly related

**Review by scarf005 — DISMISSED** [review]
| initial | after 12h | 
|:-:|:-:|
| image | image |

image

- might be better if the walls could turn into impassable fungal walls but this could be done in later PRs
- trying to load dangerous fungus mod from old BN executable silently fails and goes back to main menu; i'm not sure we should handle this case.

cc @chaosvolt, @KheirFerrum and @Coolthulhu for inputs

File: src/monattack.h (hunk 1 above)
Comment (4nonch) [reply src/monattack.h:48]: just for the sake of beauty since comments were not leveled
Screenshot_6

File: src/cached_options.h
Comment (4nonch) [reply src/cached_options.h:1]: make sense, we'll see it

**Comment by 4nonch** [comment]
> might be better if the walls could turn into impassable fungal walls but this could be done in later PRs
Actually walls could be infected but walls on screenshot not marked as "FLAMMABLE" so they not got infested at the end.
Screenshot_7
Maybe need to add check for FLAMMABLE_HARD or you meant something else like new hardened infested wall-type?
Screenshot_8
> trying to load dangerous fungus mod from old BN executable silently fails and goes back to main menu; i'm not sure we should handle this case.
Because it reference new balance option which not existed before the PR.
image

File: src/monattack.h (hunk 1 above)
Comment (scarf005) [reply src/monattack.h:48]:
```
/// Broadly scatter aerobics
bool fungus_haze( monster *z );
```
might as well just make them proper docstring in order to add IDE support

File: src/cached_options.h
Comment (4nonch) [reply src/cached_options.h:1]: Attempt: https://github.com/cataclysmbnteam/Cataclysm-BN/pull/5094/commits/13389742b607bccca856a1f57da385ae0683f992

File: src/monattack.h (hunk 1 above)
Comment (4nonch) [reply src/monattack.h]: done

File: src/cached_options.h
Comment (scarf005) [reply src/cached_options.h:1]: sorry, i meant plain struct without neither constructor nor methods.

**Review by scarf005 — APPROVED** [review]

**Pull request merged**